src/interface/control_textbox.cpp

```diff
@@ -288,70 +435,136 @@ ControlTextbox::on_key_down(const SDL_KeyboardEvent& key)
   }
   else if (key.keysym.sym == SDLK_a && m_ctrl_pressed)
   {
-    m_caret_pos = 0;
-    m_secondary_caret_pos = static_cast<int>(m_charlist.size());
-    recenter_offset();
+    set_caret_pos(std::numeric_limits<int>::max());
+    set_secondary_caret_pos(0);
+    recenter_offset(true);
     return true;
   }
-  else if (key.keysym.sym == SDLK_RETURN)
+  else if (key.keysym.sym == SDLK_z && m_ctrl_pressed)
   {
-    parse_value();
+    if (m_undo_stack.empty())
+      return true;
+
+    const auto& change = m_undo_stack.back();
+    revert_value(change.string);
+    if (parse_value(true, TextAction::UNDO))
+    {
+      m_caret = std::move(change.carets.caret);
+      m_secondary_caret = std::move(change.carets.secondary_caret);
+    }
+    m_undo_stack.pop_back();
+    return true;
+  }
+  else if (key.keysym.sym == SDLK_y && m_ctrl_pressed)
+  {
+    if (m_redo_stack.empty())
+      return true;
+
+    const auto& change = m_redo_stack.back();
+    revert_value(change.string);
+    if (parse_value(true, TextAction::REDO))
+    {
+      m_caret = std::move(change.carets.caret);
+      m_secondary_caret = std::move(change.carets.secondary_caret);
+    }
+    m_redo_stack.pop_back();
     return true;
   }
```

Comment:
This is like... 18 if else statements... Sorry but, please refactor this somehow. I did not read this with a straight face.